### PR TITLE
`EssentialType`: Implement correct essential types for bitwise binary operators `&`, `^` and `|`.

### LINKED ISSUE #786: `RULE-10-3`: Essential type of binary bitwise operations is incorrect
### Affected rules
 - `RULE-10-1`
 - `RULE-10-3`
 - `RULE-10-4`
 - `RULE-10-5`
 - `RULE-10-6`
 - `RULE-10-7`
 - `RULE-10-8`
 - `RULE-12-2`
 
### Description

The `|`, `&` and `^` binary bitwise operations are currently incorrectly assigned the standard type as the essential type of the expression. Instead, their treatment under essential types should be similar to other arithmetic expressions, in that if both operands are consistently signed, the essential type is the has the same signedness, and a rank of the larger of the two operands.

### Example

```cpp
uint8_t u8;
uint16_t u16;
u8 & u16; // Should have an essential type of unsigned short, not the standard type of int
```

### PR DESCRIPTION
**Pull request by lcartey**
## Description

Fixes #786.

Implements correct behaviour for binary bitwise operators.

## Change request type

 - [ ] Release or process automation (GitHub workflows, internal scripts)
 - [ ] Internal documentation
 - [ ] External documentation
 - [x] Query files (`.ql`, `.qll`, `.qls` or unit tests)
 - [ ] External scripts (analysis report or other code shipped as part of a release)

## Rules with added or modified queries

 - [ ] No rules added
 - [ ] Queries have been added for the following rules:
     - _rule number here_
 - [x] Queries have been modified for the following rules:
     - `RULE-10-1`
     - `RULE-10-3`
     - `RULE-10-4`
     - `RULE-10-5`
     - `RULE-10-6`
     - `RULE-10-7`
     - `RULE-10-8`
     - `RULE-12-2`
 
## Release change checklist

A change note ([development_handbook.md#change-notes](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#change-notes)) is required for any pull request which modifies:

 - The structure or layout of the release artifacts.
 - The evaluation performance (memory, execution time) of an existing query.
 - The results of an existing query in any circumstance.

If you are only adding new rule queries, a change note is not required.

_**Author:**_ Is a change note required? 
  - [x] Yes
  - [ ] No

🚨🚨🚨
_**Reviewer:**_ Confirm that format of *shared* queries (not the .qll file, the
.ql file that imports it) is valid by running them within VS Code. 
  - [ ] Confirmed

_**Reviewer:**_ Confirm that either a change note is not required or the change note is required and has been added.
  - [x] Confirmed

## Query development review checklist

For PRs that add new queries or modify existing queries, the following checklist should be completed by both the author and reviewer:

### Author

 - [x] Have all the relevant rule package description files been checked in?
 - [x] Have you verified that the metadata properties of each new query is set appropriately?
 - [x] Do all the unit tests contain both "COMPLIANT" and "NON_COMPLIANT" cases?
 - [x] Are the alert messages properly formatted and consistent with the [style guide](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#query-style-guide)?
 - [x] Have you run the queries on OpenPilot and verified that the performance and results are acceptable?<br />_As a rule of thumb, predicates specific to the query should take no more than 1 minute, and for simple queries be under 10 seconds. If this is not the case, this should be highlighted and agreed in the code review process._
 - [x] Does the query have an appropriate level of in-query comments/documentation?
 - [x] Have you considered/identified possible edge cases?
 - [x] Does the query not reinvent features in the standard library?
 - [x] Can the query be simplified further (not golfed!)

### Reviewer 

 - [x] Have all the relevant rule package description files been checked in?
 - [x] Have you verified that the metadata properties of each new query is set appropriately?
 - [x] Do all the unit tests contain both "COMPLIANT" and "NON_COMPLIANT" cases?
 - [x] Are the alert messages properly formatted and consistent with the [style guide](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#query-style-guide)?
 - [x] Have you run the queries on OpenPilot and verified that the performance and results are acceptable?<br />_As a rule of thumb, predicates specific to the query should take no more than 1 minute, and for simple queries be under 10 seconds. If this is not the case, this should be highlighted and agreed in the code review process._
 - [x] Does the query have an appropriate level of in-query comments/documentation?
 - [x] Have you considered/identified possible edge cases?
 - [x] Does the query not reinvent features in the standard library?
 - [ ] Can the query be simplified further (not golfed!)

### CODE REVIEW

File: c/misra/test/c/misra/test.c [hunk 1]
```diff
@@ -79,4 +79,122 @@ void testControlChar() {
...
+
+  s8 ^ s8;   // Essentially signed, char
+  s16 ^ s8;  // Essentially signed, short
+  s8 ^ s16;  // Essentially signed, short
+  s32 ^ s8;  // Essentially signed, int
+  s8 ^ s32;  // Essentially signed, int
+  s32 ^ s16; // Essentially signed, int
+  s16 ^ s32; // Essentially signed, int
+
+  u32 & s32; // Essentially signed, int
```
Comment (MichaelRFairhurst) [inline c/misra/test/c/misra/test.c:142]: These are essentially unsigned ints in the .expected file, which I think is correct based on c11 spec

> if the operand that has unsigned integer type has rank greater or
equal to the rank of the type of the other operand, then the operand with
signed integer type is converted to the type of the operand with unsigned
integer type.

**Review by MichaelRFairhurst — CHANGES_REQUESTED** [review]

File: c/misra/src/codingstandards/c/misra/EssentialTypes.qll [hunk 2]
```diff
@@ -353,6 +353,51 @@ class EssentialBinaryArithmeticExpr extends EssentialExpr, BinaryArithmeticOpera
   }
 }
 
+class EssentialBinaryBitwiseExpr extends EssentialExpr, BinaryBitwiseOperation {
+  EssentialBinaryBitwiseExpr() {
+    not this instanceof LShiftExpr and
+    not this instanceof RShiftExpr
```
[the remark this answers is not in the document]
Comment (lcartey) [reply c/misra/src/codingstandards/c/misra/EssentialTypes.qll:359]: Agreed.

File: c/misra/src/codingstandards/c/misra/EssentialTypes.qll [hunk 3]
```diff
@@ -353,6 +353,51 @@ class EssentialBinaryArithmeticExpr extends EssentialExpr, BinaryArithmeticOpera
...
+      rightEssentialTypeCategory = getEssentialTypeCategory(rightEssentialType)
+    |
+      if
+        leftEssentialTypeCategory = EssentiallySignedType() and
+        rightEssentialTypeCategory = EssentiallySignedType()
+      then
+        if exists(getValue())
+        then result = stlr(this)
+        else (
+          if leftEssentialType.getSize() > rightEssentialType.getSize()
```
[the remark this answers is not in the document]
Comment (lcartey) [reply c/misra/src/codingstandards/c/misra/EssentialTypes.qll:380]: Done.

File: c/misra/src/codingstandards/c/misra/EssentialTypes.qll [hunk 4]
```diff
@@ -353,6 +353,51 @@ class EssentialBinaryArithmeticExpr extends EssentialExpr, BinaryArithmeticOpera
...
+      Type leftEssentialType, Type rightEssentialType,
+      EssentialTypeCategory leftEssentialTypeCategory,
+      EssentialTypeCategory rightEssentialTypeCategory
+    |
+      leftEssentialType = getEssentialType(getLeftOperand()) and
+      rightEssentialType = getEssentialType(getRightOperand()) and
+      leftEssentialTypeCategory = getEssentialTypeCategory(leftEssentialType) and
+      rightEssentialTypeCategory = getEssentialTypeCategory(rightEssentialType)
+    |
+      if
```
[the remark this answers is not in the document]
Comment (lcartey) [reply c/misra/src/codingstandards/c/misra/EssentialTypes.qll:373]: I've avoided needing to do this by refactoring the binary operation and conditional expression classes to remove the duplication that would justify these extractions. I've also introduced a `EssentiallySignedOrUnsignedType` to avoid some additional boiler plate.

File: c/misra/test/c/misra/test.c (hunk 1 above)
Comment (lcartey) [reply c/misra/test/c/misra/test.c:142]: Thanks, fixed.

File: c/misra/src/codingstandards/c/misra/EssentialTypes.qll [hunk 5]
```diff
@@ -305,54 +316,62 @@ class EssentialBinaryArithmeticExpr extends EssentialExpr, BinaryArithmeticOpera
...
+        leftEssentialTypeCategory = EssentiallyCharacterType() and
+        rightEssentialTypeCategory instanceof EssentiallySignedOrUnsignedType
+      then result instanceof PlainCharType
+      else result = super.getEssentialType()
     )
   }
 }
 
+// }
```
Comment (MichaelRFairhurst) [inline c/misra/src/codingstandards/c/misra/EssentialTypes.qll:374]: Stray comment

**Review by MichaelRFairhurst — CHANGES_REQUESTED** [review]
Refactor is **\*chef's kiss**\* fantastic! Great work untangling.

Perfect except for the one straggling commented out closing brace.

File: c/misra/src/codingstandards/c/misra/EssentialTypes.qll (hunk 5 above)
Comment (lcartey) [reply c/misra/src/codingstandards/c/misra/EssentialTypes.qll]: Addressed.

**Review by MichaelRFairhurst — APPROVED** [review]
